src/System.Utf8String.Experimental/tests/System/Utf8SpanTests.cs

```diff
@@ -0,0 +1,352 @@
+                Assert.NotEqual(span[0..1].GetHashCode(StringComparison.CurrentCultureIgnoreCase), span[5..6].GetHashCode(StringComparison.CurrentCultureIgnoreCase)); // 'i' shouldn't match 'I'
+                Assert.Equal(Utf8Span.Empty.GetHashCode(StringComparison.CurrentCultureIgnoreCase), span[^0..].GetHashCode(StringComparison.CurrentCultureIgnoreCase)); // null should equal empty
+            });
+        }
+
+        [Theory]
+        [InlineData("", true)]
+        [InlineData("Hello", true)]
+        [InlineData("\u1234", false)]
+        public static void IsAscii(string input, bool expected)
+        {
+            using BoundedUtf8Span boundedSpan = new BoundedUtf8Span(input);
+
+            Assert.Equal(expected, boundedSpan.Span.IsAscii());
+        }
+
+        [Theory]
+        [InlineData(null, true)]
+        [InlineData("", true)]
+        [InlineData(" \u2028\u2029\t\v", true)]
+        [InlineData(" x\r\n", false)]
+        [InlineData("\r\nhello\r\n", false)]
+        [InlineData("\r\n\0\r\n", false)]
+        [InlineData("\r\n\r\n", true)]
+        public static void IsEmptyOrWhiteSpace(string input, bool expected)
+        {
+            using BoundedUtf8Span boundedSpan = new BoundedUtf8Span(input);
+
+            Assert.Equal(expected, boundedSpan.Span.IsEmptyOrWhiteSpace());
+        }
+
+        [Theory]
+        [InlineData("", "..")]
+        [InlineData("Hello", "1..")]
+        [InlineData("Hello", "1..2")]
+        [InlineData("Hello", "1..^2")]
+        [InlineData("Hello", "^2..")]
+        [InlineData("Hello", "^0..")]
+        [InlineData("résumé", "1..^2")] // include first 'é', exclude last 'é'
+        [InlineData("résumé", "^2..")] // include only last 'é'
+        public static void Indexer_Success(string input, string rangeExpression)
+        {
+            Range range = ParseRangeExpr(rangeExpression);
+
+            using BoundedUtf8Span boundedSpan = new BoundedUtf8Span(input);
+            Utf8Span originalSpan = boundedSpan.Span;
+            Utf8Span slicedSpan = originalSpan[range]; // shouldn't throw
+
+            ref byte startOfOriginalSpan = ref MemoryMarshal.GetReference(originalSpan.Bytes);
+            ref byte startOfSlicedSpan = ref MemoryMarshal.GetReference(slicedSpan.Bytes);
+
+            // Now ensure the slice was correctly produced by comparing the references directly.
+
+            (int offset, int length) = range.GetOffsetAndLength(originalSpan.Bytes.Length);
+            Assert.True(Unsafe.AreSame(ref startOfSlicedSpan, ref Unsafe.Add(ref startOfOriginalSpan, offset)));
+            Assert.Equal(length, slicedSpan.Bytes.Length);
+        }
+
+        [Theory]
+        [InlineData("résumé", "2..")] // try to split the first 'é'
+        [InlineData("résumé", "..^1")] // try to split the last 'é'
+        public static void Indexer_ThrowsIfTryToSplitMultiByteSubsequence(string input, string rangeExpression)
```

Comment:
why this doesn't take range directly instead of expression string?

Reply:
Only consts / primitives can be used as attribute data. :(
This was the most readable way I could think of to work around it.